internal/pkg/agent/cmd/container_init_linux.go

```diff
@@ -0,0 +1,313 @@
+func initContainer(streams *cli.IOStreams) (*exec.Cmd, error) {
+	isRoot, _ := utils.HasRoot()
+	if !skipFileCapabilities && !isRoot {
+		executable, err := os.Executable()
+		if err != nil {
+			return nil, err
+		}
+
+		logInfo(streams, "agent container initialisation - file capabilities")
+		updated, err := updateFileCapsFromBoundingSet(executable)
+		if err != nil {
+			return nil, err
+		}
+
+		if updated {
+			// new capabilities were added thus we need to re-exec agent to pick them up
+			var args []string
+			if len(os.Args) > 1 {
+				args = append(args, os.Args[1:]...)
+			}
+			// add skipFileCapabilitiesFlag flag to skip reapplying the file capabilities
+			args = append(args, fmt.Sprintf("--%s", skipFileCapabilitiesFlag))
+
+			cmd := exec.Command(executable, args...)
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+			cmd.Env = os.Environ()
+			return cmd, nil
+		}
+	}
+
+	if !isRoot {
+		// if we are not root, we need to raise the ambient capabilities
+		logInfo(streams, "agent container initialisation - ambient capabilities")
+		if err := raiseAmbientCapabilities(); err != nil {
+			return nil, err
+		}
+	}
+
+	// check if we have DAC_OVERRIDE
+	// Note: that even if we running under root (uid = 0), we may not have DAC_OVERRIDE which is there by default
+	// ( e.g. with cap-drop: ALL). Thus, we won't be able to read/write any file that doesn't belong to us
+	procSet := capProc()
+	hasOverride, err := procSet.GetFlag(cap.Effective, cap.DAC_OVERRIDE)
+	if err != nil {
+		return nil, err
+	}
+	if !hasOverride {
+		// we need to chown all paths
+		logInfo(streams, "agent container initialisation - chown paths")
+
+		if err = chownPaths(); err != nil {
+			return nil, err
+		}
+	}
+
+	return nil, nil
+}
+
+// raiseAmbientCapabilities will attempt to raise all capabilities present in the Effective set of the running process
+// to the Ambient set. Note that for security reasons CAP_CHOWN, CAP_SETPCAP, CAP_SETFCAP are excluded.
+func raiseAmbientCapabilities() error {
+	caps, err := getAmbientCapabilitiesFromEffectiveSet()
+	if err != nil {
+		return err
+	}
+
+	iab := cap.NewIAB()
+	for _, capVal := range caps {
+		err = iab.SetVector(cap.Inh, true, capVal)
+		if err != nil {
+			return fmt.Errorf("failed to set inheritable vector: %w", err)
+		}
+		err = iab.SetVector(cap.Amb, true, capVal)
+		if err != nil {
+			return fmt.Errorf("failed to set ambient vector: %w", err)
+		}
+	}
+
+	return iab.SetProc()
+}
+
+// getAmbientCapabilitiesFromEffectiveSet returns the capabilities that are in the Effective set of the running process
+// excluding CAP_CHOWN, CAP_SETPCAP, and CAP_SETFCAP.
+func getAmbientCapabilitiesFromEffectiveSet() ([]cap.Value, error) {
+	set := capProc()
+	var caps []cap.Value
+
+	for capVal := cap.Value(0); capVal < cap.MaxBits(); capVal++ {
+
+		switch capVal {
+		case cap.CHOWN, cap.SETPCAP, cap.SETFCAP:
+			// don't set these as they shouldn't be required by any exec'ed child process
```

Comment:
does this affect the ability for the Elastic Agent running as non-root to write a file and set its ownership to its current user? I believe there are specific code paths where it does perform chown always to ensure that the ownership is correct, will that need to be skipped?

Reply:
no these are skipped only for Ambient permissions thus only subprocesses won't have them. the agent binary has cap_chown so it will always be able to chown files

Reply:
wonder how that affects some of the subprocess that store state, like filebeat that writes a registry.

I would hope that it doesn't cause an issue.

Reply:
I have tested manually this agent image with kubernetes integration and the container logs appear just fine but I trust your gut instincts here. What do you say to allow CAP_CHOWN? CAP_SETPCAP and CAP_SETFCAP shouldn't be necessary right?

Reply:
I would say it's only the chown that I worry about. Just because it shipping the logs doesn't mean it's having issues with writing the registry. You could also just check the file system and logs to see if the registry is working correctly. 

If it doesn't seem like much of a concern to allow them to keep that ability we might want to just to be safe.

Reply:
ok CAP_CHOWN allowed then just for now to be better safe than sorry. But maybe in the future we would like to extend a bit the package spec (agent integrations) and there capture all linux capabilities needed for each integration (and beat) so we know exactly what to allow and what to filter out 🙂

Reply:
ok CAP_CHOWN is part of Ambient Set and thus given to the individual beats spawned by agent, feel free to unresolve if you spot anything